legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java

```diff
@@ -55,76 +58,66 @@ public boolean equals(Object other)
-                this.getCreationTime().equals(that.getCreationTime()) &&
-                this.tokenManager.equals(that.tokenManager);
+            Objects.equals(this.getProfile(), that.getProfile()) &&
+            this.getCreationTime().equals(that.getCreationTime()) &&
+            this.appInfo.equals(that.appInfo) &&
+            Objects.equals(this.token, that.token);
     }
 
     @Override
     public int hashCode()
     {
-        return getCreationTime().hashCode() ^ getUserId().hashCode() ^ this.tokenManager.hashCode();
-    }
-
-    @Override
-    public Set<GitLabMode> getValidModes()
-    {
-        return this.tokenManager.getValidModes();
-    }
-
-    @Override
-    public boolean isValidMode(GitLabMode mode)
-    {
-        return this.tokenManager.isValidMode(mode);
+        return getCreationTime().hashCode() ^ getUserId().hashCode() ^ this.appInfo.hashCode() ^ Objects.hashCode(this.token);
     }
 
     @Override
-    public boolean gitLabOAuthCallback(GitLabMode mode, String code)
+    public boolean gitLabOAuthCallback(String code)
     {
-        return this.tokenManager.gitLabOAuthCallback(mode, code);
+        GitLabToken token = GitLabToken.newGitLabToken(Constants.TokenType.OAUTH2_ACCESS, GitLabOAuthAuthenticator.newAuthenticator(appInfo).getOAuthTokenFromAuthCode(code));
+        return !token.equals(this.token);
     }
 
     @Override
-    public GitLabToken getGitLabToken(GitLabMode mode)
+    public GitLabToken getGitLabToken()
     {
-        return this.tokenManager.getGitLabToken(mode);
+        return this.token;
     }
 
     @Override
-    public void clearGitLabTokens()
+    public void clearGitLabToken()
     {
-        this.tokenManager.clearGitLabTokens();
+        this.token = null;
     }
 
     @Override
-    public void putGitLabToken(GitLabMode mode, GitLabToken token)
+    public void setGitLabToken(GitLabToken token)
     {
-        this.tokenManager.putGitLabToken(mode, token);
+        this.token = token;
     }
 
     @Override
-    public GitLabModeInfo getModeInfo(GitLabMode mode)
+    public GitLabAppInfo getAppInfo()
     {
-        return this.tokenManager.getModeInfo(mode);
+        return this.appInfo;
     }
 
     @Override
     public Token.TokenBuilder encode(Token.TokenBuilder builder)
     {
-        return this.tokenManager.encode(super.encode(builder));
+        return GitLabTokenHelper.encodeGitLabToken(super.encode(builder), this.token, this.appInfo);
     }
 
     @Override
     protected void writeToStringInfo(StringBuilder builder)
     {
-        this.tokenManager.appendTokenInfo(builder.append(' '));
+        GitLabTokenHelper.appendGitLabTokenInfo(builder.append(' '), this.token);
     }
 
-    private static GitLabTokenManager possiblyInitializeTokenManager(GitLabTokenManager tokenManager, OidcProfile profile)
+    private void possiblyInitializeGitLabToken(GitLabAppInfo appInfo, OidcProfile profile)
```

Comment:
```suggestion
    private static GitLabToken possiblyInitializeGitLabToken(GitLabAppInfo appInfo, OidcProfile profile)
```

Reply:
Don't think I will do this. I know it looks like we diverge from the original but since this method possibly return `null` (in case you want it to return `GitLabToken`), and we need to set its return value in the constructor, so something like 

```java
construtor(...)
{
        super(profile, userId, creationTime);
        this.appInfo = appInfo;
        this.token = token;
        GitLabToken newToken = this.possiblyInitializeGitLabToken(appInfo, profile);
         if (newToken != null)
         {
             this.token = newToken;
         }
    }
```

I think that's clunky, so there's no real benefit to say this method must return a value.

Reply:
I think you're misunderstanding the desired logic here. It should simply be:

`this.token = possiblyInitializeGitLabToken(appInfo, profile);`

The main point of possiblyInitializeGitLabToken is to check whether the token from the profile (i.e., the one from authentication) can be used in the session.  There's no guarantee that it is suitable for connecting to the GitLab API. That's what we're ensuring here. If it returns null, then we want `this.token` to be null.

Another thing is that when possible it's best to use static methods to initialize values in a constructor. It ensures there's no conflict with other state on the instance. It also makes it clear in the constructor exactly what information is being used in the initialization.

Reply:
Gotcha! Done now